File: algorithm/src/main/java/sulwish/L45.java
package sulwish;

public class L45 {
    public static void main(String[] args) {
        System.out.println(new L45().jump(new int[]{2, 3, 1, 1, 4}));
    }

    public int jump(int[] nums) {
        int len = nums.length;
        if (len == 1) return 0;
        int left = 0;
        int right = 0;
        int max = nums[left];
        int step = 0;
        while (true) {
            for (int i = left; i <= right; i++) {
                max = Math.max(max, i + nums[i]);
            }
            if (max >= len - 1) return step + 1;
            else {
                left = right + 1;
                right = max;
                step++;
            }
        }
    }
}

File: pys/pfda/L45.py
"""
给定一个长度为 n 的 0 索引整数数组 nums。初始位置为 nums[0]。

每个元素 nums[i] 表示从索引 i 向前跳转的最大长度。换句话说，如果你在 nums[i] 处，你可以跳转到任意 nums[i + j] 处:

0 <= j <= nums[i]
i + j < n
返回到达 nums[n - 1] 的最小跳跃次数。生成的测试用例可以到达 nums[n - 1]。
"""
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 1:
            return 0
        limit = min(nums[0], len(nums)-1)
        step = 0
        pos = 0
        while limit < len(nums)-1:
            step += 1
            l = limit
            while pos <= limit:
                l = max(l, pos + nums[pos])
                pos += 1
            limit = l

        return step + 1

if __name__ == '__main__':
    print(Solution().jump([1,2,3]))